src/NServiceBus.Core/Unicast/Messages/MessageMetadata.cs

```diff
@@ -16,7 +19,7 @@ public partial class MessageMetadata
         public MessageMetadata(Type messageType, IEnumerable<Type> messageHierarchy = null)
```

Comment:
`RegisterMessageType` in `MessageMetadataRegistry` creates the hierarchy. It currently calls `ToList()`. Change it `ToArray()` and make it in the ctor `Type[]`. We could even think about moving the `emptyHierarchy` to the registry as well and remove the null check here